File: C_Alghorithms/C_Alghorithms/two_way_merge.cpp
#include <iostream>
#include <vector>
#include <cassert>
#include <algorithm>

class TwoWayMerger {
public:
    // Merges two sorted vectors into one sorted vector
    std::vector<int> merge_two_lists(const std::vector<int>& list1, const std::vector<int>& list2) {
        std::vector<int> merged_list;
        int i = 0, j = 0;

        // Merge until one list is exhausted
        while (i < list1.size() && j < list2.size()) {
            if (list1[i] < list2[j]) {
                merged_list.push_back(list1[i++]);
            }
            else {
                merged_list.push_back(list2[j++]);
            }
        }

        // Add remaining elements from either list
        while (i < list1.size()) {
            merged_list.push_back(list1[i++]);
        }

        while (j < list2.size()) {
            merged_list.push_back(list2[j++]);
        }

        return merged_list;
    }

    // Performs one step of merging in pairs for a list of sorted lists
    std::vector<std::vector<int>> one_step_k_way_merge(const std::vector<std::vector<int>>& list_of_lists) {
        if (list_of_lists.size() <= 1) {
            return list_of_lists;
        }

        std::vector<std::vector<int>> merged_lists;
        for (size_t i = 0; i < list_of_lists.size(); i += 2) {
            if (i + 1 < list_of_lists.size()) {
                merged_lists.push_back(merge_two_lists(list_of_lists[i], list_of_lists[i + 1]));
            }
            else {
                merged_lists.push_back(list_of_lists[i]);
            }
        }

        return merged_lists;
    }

    // Recursively merges a list of sorted lists into one sorted list
    std::vector<int> k_way_merge(std::vector<std::vector<int>> list_of_lists) {
        if (list_of_lists.size() == 1) {
            return list_of_lists[0];
        }

        std::vector<std::vector<int>> merged_lists = one_step_k_way_merge(list_of_lists);
        return k_way_merge(merged_lists);
    }
};

class TestMerger {
public:
    static void run_tests() {
        TwoWayMerger merger;

        // Test case 1
        std::vector<std::vector<int>> test_case1 = { {1, 2, 3}, {4, 5, 7}, {-2, 0, 6}, {5} };
        std::vector<int> result = merger.k_way_merge(test_case1);
        std::vector<int> expected1 = { -2, 0, 1, 2, 3, 4, 5, 5, 6, 7 };
        assert(result == expected1 && "Test Case 1 Failed");

        // Test case 2
        std::vector<std::vector<int>> test_case2 = { {-2, 4, 5, 8}, {0, 1, 2}, {-1, 3, 6, 7} };
        result = merger.k_way_merge(test_case2);
        std::vector<int> expected2 = { -2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8 };
        assert(result == expected2 && "Test Case 2 Failed");

        // Test case 3
        std::vector<std::vector<int>> test_case3 = { {-1, 1, 2, 3, 4, 5} };
        result = merger.k_way_merge(test_case3);
        std::vector<int> expected3 = { -1, 1, 2, 3, 4, 5 };
        assert(result == expected3 && "Test Case 3 Failed");

        std::cout << "Two Way Merge Sort: All tests passed!\n";
    }
};

File: C_Alghorithms/python/sorting_algorithms/two_way_merge.py
# Two way sort algorithm
# Based on Coursera Algorithms for searching, sorting and Indexing

class TwoWayMerger(object):
    def merge_two_lists(self, list1, list2):
        """Merges two sorted lists into one sorted list."""
        merged_list = []
        i, j = 0, 0

        while i < len(list1) and j < len(list2):
            if list1[i] < list2[j]:
                merged_list.append(list1[i])
                i += 1
            else:
                merged_list.append(list2[j])
                j += 1

        # Add remaining elements from either list
        merged_list.extend(list1[i:])
        merged_list.extend(list2[j:])

        return merged_list

    def one_step_k_way_merge(self, list_of_lists):
        """Performs one step of merging in pairs for a list of sorted lists."""
        if len(list_of_lists) <= 1:
            return list_of_lists

        merged_lists = []
        for i in range(0, len(list_of_lists), 2):
            if i + 1 < len(list_of_lists):
                merged_lists.append(self.merge_two_lists(list_of_lists[i], list_of_lists[i + 1]))
            else:
                merged_lists.append(list_of_lists[i])

        return merged_lists

    def k_way_merge(self, list_of_lists):
        """Recursively merges a list of sorted lists into one sorted list."""
        if len(list_of_lists) == 1:
            return list_of_lists[0]

        merged_lists = self.one_step_k_way_merge(list_of_lists)
        return self.k_way_merge(merged_lists)


class TestMerger:

    @staticmethod
    def run_tests():
        merger = TwoWayMerger()

        # Test case 1
        result = merger.k_way_merge([[1, 2, 3], [4, 5, 7], [-2, 0, 6], [5]])
        assert result == [-2, 0, 1, 2, 3, 4, 5, 5, 6, 7], "Test Case 1 Failed"

        # Test case 2
        result = merger.k_way_merge([[-2, 4, 5, 8], [0, 1, 2], [-1, 3, 6, 7]])
        assert result == [-2, -1, 0, 1, 2, 3, 4, 5, 6, 7, 8], "Test Case 2 Failed"

        # Test case 3
        result = merger.k_way_merge([[-1, 1, 2, 3, 4, 5]])
        assert result == [-1, 1, 2, 3, 4, 5], "Test Case 3 Failed"

        print('Two way merge: All tests passed!')


if __name__ == "__main__":
    TestMerger.run_tests()
